### DISCUSSION TITLE
What environments is Mosaic envisioned to run in?

### DISCUSSION THREAD

**Discussion by mhkeller** [uwdata/mosaic#156] [General]
This looks like a great project! 

I have a question on what kinds of workflows this could be integrated with. I see that you have examples for Jupyter and Observable — and the documentation talks about running DuckDB-wasm [from inside a browser](https://uwdata.github.io/mosaic/get-started/) like this and then handling the HTML insertion on your own:

```js
import * as vg from "@uwdata/vgplot";

// configure the coordinator to use DuckDB-WASM
// creates a new database instance running in-browser
const wasm = await vg.wasmConnector();
vg.coordinator().databaseConnector(wasm);

// load data into the database
// executes a query generated by the loadCSV helper
vg.coordinator().exec(vg.loadCSV("stocks", "stock-data.csv"));

// create an area chart, returned as an HTML element
// you can subsequently add this to your webpage
const chart = vg.plot(
  vg.areaY(vg.from("stocks"), { x: "Date", y: "Price" })
);
```

Do you plan or have you considered the ability to run similar code from a Node environment and viewing the charts from VS Code or other editors? I would find it super useful to use Arquero and Mosaic to visualize analysis from a local CSV or connect DuckDB to postgres. The R package [httpgd](https://nx10.github.io/httpgd/articles/b01_vscode.html) is a good example of integrating plots into VS Code (side note: being able to write the plots to file would also be really useful). 

Thanks for your work on this and very excited to see how it develops.

**Comment by domoritz** [comment]
Mosaic already supports running queries in DuckDB running in WebAssembly and well as [Node (connected via REST or WebSockets)](https://github.com/uwdata/mosaic/blob/5b8a8b2e2fc1ce6bf1f3a1566cfb62459660f9b6/packages/duckdb/src/data-server.js). The visualizations in vgplot are currently implanted to run in a browser. It should be fairly straight forward to integrate Mosaic into VSCode or other web-based environments. I think this could be done similar to the Jupyter widget (which also works in Google Colab). I like the idea to extend to more platforms but we don't have any immediate plans to implement e.g. a VSCode plugin.

**Reply by mhkeller** [reply]
Thanks for the info. If you're ever able to spend the time on it, I'll add a +1 for a plugin. I could try implementing it but I imagine there may be some intricacies I'm not aware of.

Is the main use case you're picturing to be someone working within Observable or be coding inside of script tags on a local HTML file that they are then viewing via a local server?

**Reply by domoritz** [reply]
Awesome. Mosaic in its current form is providing the infrastructure to build interactive linked charts. Right now, we mainly target developers but as we develop Mosaic I can imagine that we will make it easier for people to build dashboards on various platforms without the need to write the boilerplate to set up Mosaic there.
